In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
import os
print(os.getcwd())

C:\Users\Dell\OneDrive\Desktop\RetailHub-India-Business-Intelligence\05_Python


In [3]:
import os

print(os.listdir(".."))

['.git', '.gitignore', '01_business_requirements', '02_raw_data', '03_cleaned_data', '04_SQL', '05_Python', '06_Power_BI', '07_report', '08_Presentation', '09_Images', '10_Dataset_dictionary', 'README.md']


In [4]:
customer=pd.read_csv("../02_raw_data/customers.csv")
order_items=pd.read_csv("../02_raw_data/order_line_items.csv")
orders=pd.read_csv("../02_raw_data/orders.csv")
catalog=pd.read_csv("../02_raw_data/sku_catalog.csv")
sessions=pd.read_csv("../02_raw_data/website_sessions.csv")

## Data Inventory ##
In this step, we inspect the size of each dataset to understand the amount of data available for analysis.

In [5]:
customer.shape

(16865, 11)

In [6]:
orders.shape

(30000, 16)

In [7]:
order_items.shape

(42121, 10)

In [8]:
catalog.shape

(55, 5)

In [9]:
sessions.shape

(475658, 14)

## Understanding the Columns

In this step, we inspect the columns of each table to understand what business information each dataset contains.

In [10]:
customer.columns

Index(['Customer ID', 'Name', 'First order date', 'Total orders',
       'Total revenue', 'Average order value', 'Time to 2nd purchase',
       'Last purchase date', 'City / tier',
       'Acquisition channel (first touch)', 'RePurchased'],
      dtype='object')

In [11]:
customer.sample(10)

,Customer ID,Name,First order date,Total orders,Total revenue,Average order value,Time to 2nd purchase,Last purchase date,City / tier,Acquisition channel (first touch),RePurchased
8869,CUST-0014200,Maya Radhakrishnan,07-05-2022,1,1689,1689.00,NaN,07-05-2022,Delhi / Tier 1,Meta,N
10295,CUST-0016498,Azaan Chana,13-01-2023,1,2656,2656.00,NaN,13-01-2023,Mumbai / Tier 1,Meta,N
5553,CUST-0008932,Hema Devi,23-12-2024,1,2147,2147.00,NaN,23-12-2024,Mysuru / Tier 2,Offline (Friends/Family),N
2535,CUST-0004116,Pavani Subramaniam,21-12-2025,1,3239,3239.00,NaN,21-12-2025,Mumbai / Tier 1,Google,N
6804,CUST-0010923,Geetika Morar,16-07-2024,1,2681,2681.00,NaN,16-07-2024,Rajkot / Tier 2,Meta,N
732,CUST-0001212,Krishna Halder,10-06-2022,7,16291,2327.29,210.0,28-04-2025,Mumbai / Tier 1,Google,Y
15636,CUST-0024981,Forum Sundaram,22-01-2024,2,6036,3018.00,120.0,21-05-2024,Nashik / Tier 2,Meta,Y
2632,CUST-0004260,Janya Nayar,27-07-2023,3,8246,2748.67,493.0,07-08-2025,Mumbai / Tier 1,Offline (Friends/Family),Y
13242,CUST-0021161,Rachit Solanki,10-06-2023,4,9709,2427.25,74.0,23-12-2024,Kanpur / Tier 2,Meta,Y
14380,CUST-0022932,Rajeshri Nori,03-02-2024,3,8036,2678.67,83.0,09-06-2024,Vadodara / Tier 2,Meta,Y


In [12]:
orders.columns

Index(['Order ID', 'Customer ID', 'Order date & time', 'Product',
       'Order value (gross, net)', 'Order value (gross)', 'Order value (net)',
       'Discount applied (₹ + %)', 'Discount applied (₹)',
       'Discount applied (%)', 'Payment mode', 'Shipping city', 'Pincode',
       'First order vs repeat', 'Channel source (last touch)',
       'Delivered / Returned / RTO'],
      dtype='object')

In [13]:
orders.sample(5)

,Order ID,Customer ID,Order date & time,Product,"Order value (gross, net)",Order value (gross),Order value (net),Discount applied (₹ + %),Discount applied (₹),Discount applied (%),Payment mode,Shipping city,Pincode,First order vs repeat,Channel source (last touch),Delivered / Returned / RTO
18242,ORD-20240801-018243,CUST-0001450,01-08-2024 22:00,2-item cart,"3,87,43,211",3874,3211,"663,17.1%",663,17.10%,UPI,Mumbai,217362,Repeat,Google,Delivered
28530,ORD-20251030-028531,CUST-0005103,30-10-2025 10:15,Mixed Cart,"3,34,82,577",3348,2577,"770,23.0%",770,23.00%,Card,Bengaluru,540976,Repeat,Meta,Delivered
14669,ORD-20240208-014670,CUST-0007915,08-02-2024 13:03,2-item cart,"3,23,42,651",3234,2651,"583,18.0%",583,18.00%,Card,Patna,361397,First,Offline (Friends/Family),Delivered
3078,ORD-20220721-003079,CUST-0009407,21-07-2022 12:28,Mixed Cart,"2,50,02,270",2500,2270,"231,9.2%",231,9.20%,Card,Mysuru,824137,Repeat,Google,Returned
13604,ORD-20231220-013605,CUST-0021075,20-12-2023 14:02,Mixed Cart,"2,77,62,386",2776,2386,"391,14.1%",391,14.10%,UPI,Nagpur,640600,Repeat,Meta,Delivered


In [14]:
order_items.columns

Index(['Order ID', 'SKU ID', 'Category (top, bottom, outerwear, etc.)', 'Size',
       'Color', 'MRP', 'Selling price', 'Discount %', 'Returned? (Y/N)',
       'Return reason (if any)'],
      dtype='object')

In [15]:
order_items.sample(5)

,Order ID,SKU ID,"Category (top, bottom, outerwear, etc.)",Size,Color,MRP,Selling price,Discount %,Returned? (Y/N),Return reason (if any)
15063,ORD-20230825-010706,SKU-CRO-0007,croptops,M,Blue,1899,2309,0.0,Y,Different than images
35739,ORD-20250616-025492,SKU-SHI-0001,shirts,XXL,Maroon,1999,1955,2.2,N,NaN
24360,ORD-20240608-017373,SKU-SHI-0001,shirts,XL,Black,1999,1020,49.0,Y,Size issue
18078,ORD-20231112-012862,SKU-JAC-0046,jackets,XS,White,4799,2578,46.3,Y,RTO
14020,ORD-20230728-009980,SKU-SHI-0015,shirts,XXL,Maroon,2199,1056,50.0,N,NaN


In [16]:
catalog.columns

Index(['SKU', 'Category', 'Vendor', 'MRP', 'Cost_per_unit'], dtype='object')

In [17]:
catalog.head(10)

,SKU,Category,Vendor,MRP,Cost_per_unit
0,SKU-SHI-0001,shirts,UrbanWeave Co.,1999,705
1,SKU-PAN-0002,pants,Apex Fashions,2699,982
2,SKU-SHI-0003,shirts,KnitCraft Studio,2799,1004
3,SKU-TSH-0004,tshirts,StitchWorks India,1499,539
4,SKU-CRO-0005,croptops,StitchWorks India,2399,846
5,SKU-SHI-0006,shirts,KnitCraft Studio,2799,1097
6,SKU-CRO-0007,croptops,Apex Fashions,1899,818
7,SKU-PAN-0008,pants,StitchWorks India,2999,1152
8,SKU-PAN-0009,pants,KnitCraft Studio,2799,1008
9,SKU-TSH-0010,tshirts,WestCo Garments,1799,782


In [18]:
sessions.columns

Index(['session_id', 'date', 'traffic_source', 'campaign_name',
       'device_category', 'city', 'sessions', 'product_views', 'add_to_cart',
       'begin_checkout', 'purchased', 'order_id', 'customer_id', 'revenue'],
      dtype='object')

In [19]:
sessions.sample(10)

,session_id,date,traffic_source,campaign_name,device_category,city,sessions,product_views,add_to_cart,begin_checkout,purchased,order_id,customer_id,revenue
398012,SESS-20250518-IN-MO-170863,18-05-2025,Influencer,NaN,mobile,Mumbai,1,4,0,0,0,NaN,NaN,0
222968,SESS-20231225-GO-TA-645841,25-12-2023,Google,NaN,tablet,Ahmedabad,1,4,0,0,0,NaN,NaN,0
42433,SESS-20220605-ME-MO-128853,05-06-2022,Meta,KS_AlwaysOn,mobile,Mumbai,1,4,0,0,0,NaN,NaN,0
46940,SESS-20220624-IN-MO-248926,24-06-2022,Influencer,NaN,mobile,Lucknow,1,1,0,0,0,NaN,NaN,0
441329,SESS-20251004-ME-MO-751484,04-10-2025,Meta,KS_Oct25_Camp_47,mobile,Pune,1,7,0,0,0,NaN,NaN,0
338700,SESS-20241128-GO-MO-258408,28-11-2024,Google,NaN,mobile,Lucknow,1,1,0,0,0,NaN,NaN,0
101119,SESS-20230103-ME-MO-726107,03-01-2023,Meta,KS_Jan23_Camp_01,mobile,Hyderabad,1,7,0,0,0,NaN,NaN,0
295177,SESS-20240802-ME-DE-205380,02-08-2024,Meta,KS_Aug24_Camp_29,desktop,Bengaluru,1,7,0,0,0,NaN,NaN,0
458780,SESS-20251111-ME-MO-294352,11-11-2025,Meta,KS_Nov25_Camp_50,mobile,Mumbai,1,3,0,0,0,NaN,NaN,0
279187,SESS-20240608-IN-MO-413071,08-06-2024,Influencer,NaN,mobile,Mumbai,1,1,0,0,0,NaN,NaN,0


## Data Types

In this step, we inspect the data types of each column to identify whether they are stored correctly for analysis.

In [20]:
customer.info()
orders.info()
order_items.info()
catalog.info()
sessions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16865 entries, 0 to 16864
Data columns (total 11 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Customer ID                        16865 non-null  object 
 1   Name                               16865 non-null  object 
 2   First order date                   16865 non-null  object 
 3   Total orders                       16865 non-null  int64  
 4   Total revenue                      16865 non-null  int64  
 5   Average order value                16865 non-null  float64
 6   Time to 2nd purchase               7748 non-null   float64
 7   Last purchase date                 16865 non-null  object 
 8   City / tier                        16865 non-null  object 
 9   Acquisition channel (first touch)  16865 non-null  object 
 10  RePurchased                        16865 non-null  object 
dtypes: float64(2), int64(2), object(7)
memory usage: 1.4+ 

## Missing Value Analysis

In this step, we identify missing values in each table to understand data completeness and determine whether the missing values are valid business cases or data quality issues.


In [21]:
customer.isnull().sum().sort_values(ascending=False)

Time to 2nd purchase                 9117
Customer ID                             0
Name                                    0
First order date                        0
Total orders                            0
Total revenue                           0
Average order value                     0
Last purchase date                      0
City / tier                             0
Acquisition channel (first touch)       0
RePurchased                             0
dtype: int64

In [22]:
orders.isnull().sum().sort_values(ascending=False)

Order ID                       0
Customer ID                    0
Order date & time              0
Product                        0
Order value (gross, net)       0
Order value (gross)            0
Order value (net)              0
Discount applied (₹ + %)       0
Discount applied (₹)           0
Discount applied (%)           0
Payment mode                   0
Shipping city                  0
Pincode                        0
First order vs repeat          0
Channel source (last touch)    0
Delivered / Returned / RTO     0
dtype: int64

In [23]:
order_items.isnull().sum().sort_values(ascending=False)

Return reason (if any)                     27244
Order ID                                       0
SKU ID                                         0
Category (top, bottom, outerwear, etc.)        0
Size                                           0
Color                                          0
MRP                                            0
Selling price                                  0
Discount %                                     0
Returned? (Y/N)                                0
dtype: int64

In [24]:
catalog.isnull().sum().sort_values(ascending=False)

SKU              0
Category         0
Vendor           0
MRP              0
Cost_per_unit    0
dtype: int64

In [25]:
sessions.isnull().sum().sort_values(ascending=False)

order_id           450894
customer_id        450894
campaign_name      247466
session_id              0
date                    0
traffic_source          0
device_category         0
city                    0
sessions                0
product_views           0
add_to_cart             0
begin_checkout          0
purchased               0
revenue                 0
dtype: int64

## Data type fix 
in this step, we will fix the dates column to data type date and other numeric values to float or int 

In [26]:
customer['First order date']=pd.to_datetime(customer['First order date'],dayfirst="True")

In [27]:
orders["Order date & time"]=pd.to_datetime(orders["Order date & time"],dayfirst="True")

In [28]:
orders["Discount applied (%)"]=(orders["Discount applied (%)"].str.replace("%","").astype(float))

In [29]:
customer['Last purchase date']=pd.to_datetime(customer['Last purchase date'],dayfirst="True")

In [30]:
sessions["date"]=pd.to_datetime(sessions["date"],dayfirst="True")

In [31]:
sessions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 475658 entries, 0 to 475657
Data columns (total 14 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   session_id       475658 non-null  object        
 1   date             475658 non-null  datetime64[ns]
 2   traffic_source   475658 non-null  object        
 3   campaign_name    228192 non-null  object        
 4   device_category  475658 non-null  object        
 5   city             475658 non-null  object        
 6   sessions         475658 non-null  int64         
 7   product_views    475658 non-null  int64         
 8   add_to_cart      475658 non-null  int64         
 9   begin_checkout   475658 non-null  int64         
 10  purchased        475658 non-null  int64         
 11  order_id         24764 non-null   object        
 12  customer_id      24764 non-null   object        
 13  revenue          475658 non-null  int64         
dtypes: datetime64[ns](1)

## Duplicate Check

Duplicate records can lead to incorrect analysis by inflating sales, customers, or orders.

In this step, we will identify duplicate rows in each table and decide whether they should be removed.

In [32]:
print(customer.duplicated().sum())

0


In [33]:
print(orders.duplicated().sum())

0


In [34]:
print(order_items.duplicated().sum())

13


In [35]:
print(catalog.duplicated().sum())

0


In [36]:
print(sessions.duplicated().sum())

2


## removing duplicates

In [ ]:
order_items[order_items.duplicated(keep=False)]

In [38]:
order_items.drop_duplicates(inplace=True)

In [39]:
sessions[sessions.duplicated(keep=False)]

,session_id,date,traffic_source,campaign_name,device_category,city,sessions,product_views,add_to_cart,begin_checkout,purchased,order_id,customer_id,revenue
44733,SESS-20220615-ME-MO-755749,2022-06-15,Meta,KS_AlwaysOn,mobile,Pune,1,4,0,0,0,NaN,NaN,0
44760,SESS-20220615-ME-MO-755749,2022-06-15,Meta,KS_AlwaysOn,mobile,Pune,1,4,0,0,0,NaN,NaN,0
322983,SESS-20241019-ME-MO-409533,2024-10-19,Meta,KS_Oct24_Camp_31,mobile,Surat,1,1,0,0,0,NaN,NaN,0
323005,SESS-20241019-ME-MO-409533,2024-10-19,Meta,KS_Oct24_Camp_31,mobile,Surat,1,1,0,0,0,NaN,NaN,0


In [40]:
sessions.drop_duplicates(inplace=True)

## Removing unwanted columns

In [41]:
orders.drop("Order value (gross, net)",axis=1,inplace=True)

In [42]:
orders.drop("Discount applied (₹ + %)",axis=1,inplace=True)

In [43]:
orders.columns

Index(['Order ID', 'Customer ID', 'Order date & time', 'Product',
       'Order value (gross)', 'Order value (net)', 'Discount applied (₹)',
       'Discount applied (%)', 'Payment mode', 'Shipping city', 'Pincode',
       'First order vs repeat', 'Channel source (last touch)',
       'Delivered / Returned / RTO'],
      dtype='object')

## Spliting columns 

In [44]:
customer[["city","tier"]]=customer["City / tier"].str.split("/",expand=True)

In [45]:
customer.drop("City / tier",axis=1,inplace=True)

In [46]:
customer.columns

Index(['Customer ID', 'Name', 'First order date', 'Total orders',
       'Total revenue', 'Average order value', 'Time to 2nd purchase',
       'Last purchase date', 'Acquisition channel (first touch)',
       'RePurchased', 'city', 'tier'],
      dtype='object')

In [47]:
customer[["city","tier"]].tail()

,city,tier
16860,Pune,Tier 1
16861,Hyderabad,Tier 1
16862,Vadodara,Tier 2
16863,Bhopal,Tier 2
16864,Delhi,Tier 1


## Investigating the purchased vs orders difference 

In [48]:
purchased_flagged = (sessions["purchased"] == 1).sum()

In [49]:
print("purchased = 1:", purchased_flagged)

purchased = 1: 29983


In [50]:
has_order_id=sessions["order_id"].notnull().sum()

In [51]:
print("order_id_not_null",has_order_id)

order_id_not_null 24764


In [52]:
mismatch = sessions[(sessions["purchased"] == 1) & (sessions["order_id"].isnull())]
print("Rows with purchased=1 but no order_id:", len(mismatch))
print("Revenue on these rows (should be 0 if truly not purchased):", mismatch["revenue"].sum())
mismatch[["session_id", "traffic_source", "purchased", "order_id", "revenue"]].head()

Rows with purchased=1 but no order_id: 5219
Revenue on these rows (should be 0 if truly not purchased): 0


,session_id,traffic_source,purchased,order_id,revenue
241,SESS-20220101-OR-MO-891977,Organic Instagram,1,NaN,0
242,SESS-20220101-OR-MO-386930,Organic Instagram,1,NaN,0
243,SESS-20220101-OR-MO-170254,Organic Instagram,1,NaN,0
257,SESS-20220101-OR-DE-809468,Organic Instagram,1,NaN,0
535,SESS-20220102-OR-MO-195094,Organic Instagram,1,NaN,0


**Finding:** All 5,219 mismatched rows have `revenue = 0`, meaning they don't distort revenue totals. This looks like a tracking gap — the session was flagged as "purchased" (e.g. reached a thank-you page) but the order wasn't successfully linked back (possibly a failed payment retry, or a tracking bug). 

## Incosistency check
In this step we will check the typo errors in the columns for alphabetical corrections and 

In [53]:
customer["Acquisition channel (first touch)"].unique()

array(['Meta', 'Google', 'Offline (Friends/Family)', 'Influencer',
       'Organic Instagram', 'Instagram DM', 'Email/SMS'], dtype=object)

In [54]:
customer["city"].unique()

array(['Mumbai ', 'Bengaluru ', 'Hyderabad ', 'Delhi ', 'Nashik ',
       'Noida ', 'Surat ', 'Chennai ', 'Rajkot ', 'Ahmedabad ', 'Patna ',
       'Lucknow ', 'Dehradun ', 'Indore ', 'Pune ', 'Kochi ', 'Kanpur ',
       'Coimbatore ', 'Nagpur ', 'Ludhiana ', 'Jaipur ', 'Kolkata ',
       'Mysuru ', 'Bhopal ', 'Gurgaon ', 'Bhubaneswar ', 'Visakhapatnam ',
       'Vadodara ', 'Ranchi ', 'Chandigarh '], dtype=object)

In [55]:
order_items["Color"].unique()

array(['Maroon', 'Olive', 'Black', 'Navy', 'Beige', 'Grey', 'White',
       'Blue'], dtype=object)

In [56]:
order_items["Category (top, bottom, outerwear, etc.)"].unique()

array(['pants', 'shirts', 'croptops', 'tshirts', 'shorts', 'jackets'],
      dtype=object)

In [57]:
orders["Payment mode"].unique()

array(['Delivered', 'RTO', 'Returned'], dtype=object)

In [58]:
orders["Shipping city"].unique()

array(['Lucknow', 'Delhi', 'Chandigarh', 'Indore', 'Mumbai',
       'Visakhapatnam', 'Dehradun', 'Kanpur', 'Surat', 'Bengaluru',
       'Pune', 'Bhubaneswar', 'Mysuru', 'Nagpur', 'Nashik', 'Kochi',
       'Patna', 'Ahmedabad', 'Rajkot', 'Bhopal', 'Hyderabad', 'Ranchi',
       'Jaipur', 'Noida', 'Chennai', 'Ludhiana', 'Vadodara', 'Coimbatore',
       'Kolkata', 'Gurgaon'], dtype=object)

In [59]:
orders["First order vs repeat"].unique()

array(['First', 'Repeat'], dtype=object)

In [60]:
orders["Channel source (last touch)"].unique()

array(['Meta', 'Organic Instagram', 'Influencer', 'Google', 'Email/SMS',
       'Offline (Friends/Family)', 'Instagram DM'], dtype=object)

In [61]:
orders["Delivered / Returned / RTO"].unique()

array(['Delivered', 'RTO', 'Returned'], dtype=object)

In [63]:
catalog["Category"].unique()

array(['shirts', 'pants', 'tshirts', 'croptops', 'shorts', 'jackets'],
      dtype=object)

In [64]:
catalog["Vendor"].unique()

array(['UrbanWeave Co.', 'Apex Fashions', 'KnitCraft Studio',
       'StitchWorks India', 'WestCo Garments'], dtype=object)

sessions["traffic_source"].unique()

In [69]:
sessions["campaign_name"].unique()

array(['KS_AlwaysOn', nan, 'KS_Jan23_Camp_02', 'KS_Jan23_Camp_01',
       'KS_Mar23_Camp_06', 'KS_Mar23_Camp_09', 'KS_Mar23_Camp_05',
       'KS_Mar23_Camp_04', 'KS_Mar23_Camp_08', 'KS_Mar23_Camp_07',
       'KS_Mar23_Camp_03', 'KS_Apr23_Camp_11', 'KS_Apr23_Camp_12',
       'KS_Apr23_Camp_10', 'KS_May23_Camp_13', 'KS_Jul23_Camp_14',
       'KS_Aug23_Camp_15', 'KS_Sep23_Camp_17', 'KS_Sep23_Camp_18',
       'KS_Sep23_Camp_16', 'KS_Oct23_Camp_19', 'KS_Oct23_Camp_20',
       'KS_Nov23_Camp_21', 'KS_Dec23_Camp_22', 'KS_Mar24_Camp_23',
       'KS_Apr24_Camp_24', 'KS_Jun24_Camp_25', 'KS_Jul24_Camp_27',
       'KS_Jul24_Camp_28', 'KS_Jul24_Camp_26', 'KS_Aug24_Camp_29',
       'KS_Sep24_Camp_30', 'KS_Oct24_Camp_32', 'KS_Oct24_Camp_31',
       'KS_Dec24_Camp_34', 'KS_Dec24_Camp_33', 'KS_Feb25_Camp_35',
       'KS_Mar25_Camp_36', 'KS_Mar25_Camp_37', 'KS_Mar25_Camp_38',
       'KS_Apr25_Camp_39', 'KS_May25_Camp_40', 'KS_Jun25_Camp_42',
       'KS_Jun25_Camp_41', 'KS_Jul25_Camp_43', 'KS_Sep25_Camp_

In [70]:
sessions["device_category"].unique()

array(['mobile', 'desktop', 'tablet'], dtype=object)

In [71]:
sessions["city"].unique()

array(['Pune', 'Mumbai', 'Bengaluru', 'Ahmedabad', 'Jaipur', 'Surat',
       'Delhi', 'Hyderabad', 'Lucknow', 'Indore'], dtype=object)

## Business Validation
Do the numbers and relationships make business sense?

In [ ]:
order_items[order_items["Selling price"] > order_items["MRP"]]

In [ ]:
order_items[
    order_items["Selling price"] > order_items["MRP"]
][["SKU ID", "MRP", "Selling price", "Discount %"]].head(20)

In [ ]:
order_items[
    order_items["Selling price"] > order_items["MRP"]
][["MRP", "Selling price", "Discount %"]].describe()

In [ ]:
order_items[["SKU ID", "MRP", "Selling price"]].head(20)

In [ ]:
catalog[["SKU","MRP"]]

In [88]:
order_items['Actual discount %'] = (order_items['MRP'] - order_items['Selling price']) 
/ order_items['MRP'] * 100 
#added a new column for actual discount 

In [92]:
order_items["Actual discount %"].describe()

count    42108.000000
mean        17.407107
std         41.205048
min       -179.853235
25%         -6.307996
50%         30.605530
75%         50.025013
max         79.183163
Name: Actual discount %, dtype: float64

## Finding: `Discount %` unreliable for 12,019 rows (Selling Price > MRP)
These rows show a markup (selling price above MRP), but `Discount %` is floored at 0 instead of negative. Kept `Selling price`/`MRP` as-is, added `Actual discount %` with the correct formula for analysis, and did not drop these rows (~28% of line items, likely real demand-based pricing).

In [ ]:
customer_revenue = orders.groupby("Customer ID")["Order value (net)"].sum()
customer_revenue

In [96]:
comparison = customer[["Customer ID", "Total revenue"]].merge(
    customer_revenue,
    on="Customer ID",
    how="left"
)

In [98]:
comparison["Difference"] = (
    comparison["Total revenue"] -
    comparison["Order value (net)"]
)

In [99]:
comparison["Difference"].describe()

count    16865.0
mean         0.0
std          0.0
min          0.0
25%          0.0
50%          0.0
75%          0.0
max          0.0
Name: Difference, dtype: float64

In [100]:
catalog[catalog['Cost_per_unit'] >= catalog['MRP']]

,SKU,Category,Vendor,MRP,Cost_per_unit


In [101]:
order_counts = orders.groupby("Customer ID")["First order vs repeat"].value_counts().unstack(fill_value=0)

In [102]:
check = customer[["Customer ID", "Total orders", "RePurchased"]].merge(
    order_counts,
    on="Customer ID",
    how="left"
)

In [109]:
check["orders_sum"] = check["First"] + check["Repeat"]

# Contradiction A: total orders in customers.csv doesn't match count of orders in orders.csv
contradiction_a = check[check["Total orders"] != check["orders_sum"]]

# Contradiction B: RePurchased = 'Y' but no orders marked 'Repeat'
contradiction_b = check[(check["RePurchased"] == "Y") & (check["Repeat"] == 0)]

# Contradiction C: Total orders > 1 but zero orders marked 'Repeat'
contradiction_c = check[(check["Total orders"] > 1) & (check["Repeat"] == 0)]

# Contradiction D: Total orders > 1 but RePurchased is No'
contradiction_d = check[(check["Total orders"] > 1) & (check["RePurchased"] == "N")]

print("Contradiction A (order count mismatch):", len(contradiction_a))
print("Contradiction B (RePurchased=Y but no Repeat orders):", len(contradiction_b))
print("Contradiction C (>1 order but no Repeat orders):", len(contradiction_c))
print("Total orders > 1 but RePurchased is No:",len(contradiction_d))

Contradiction A (order count mismatch): 0
Contradiction B (RePurchased=Y but no Repeat orders): 18
Contradiction C (>1 order but no Repeat orders): 18
Total orders > 1 but RePurchased is No: 0


In [104]:
set(contradiction_b["Customer ID"]) == set(contradiction_c["Customer ID"])

True

In [ ]:
suspect_ids = contradiction_b["Customer ID"].tolist()
orders[orders["Customer ID"].isin(suspect_ids)].sort_values(["Customer ID", "Order date & time"])

In [110]:
orders["order_rank"] = orders.groupby("Customer ID")["Order date & time"].rank(method="first")
orders["Corrected order type"] = orders["order_rank"].apply(lambda r: "First" if r == 1 else "Repeat")

In [111]:
mismatches = orders[orders["First order vs repeat"] != orders["Corrected order type"]]
print("Total mismatched rows across entire dataset:", len(mismatches))
print("Unique customers affected:", mismatches["Customer ID"].nunique())

Total mismatched rows across entire dataset: 30
Unique customers affected: 30


In [112]:
set(suspect_ids).issubset(set(mismatches["Customer ID"]))

True

In [113]:
mismatches.groupby("Customer ID")["Order date & time"].agg(lambda x: x.max() - x.min()).describe()

count                 30
mean     0 days 00:00:00
std      0 days 00:00:00
min      0 days 00:00:00
25%      0 days 00:00:00
50%      0 days 00:00:00
75%      0 days 00:00:00
max      0 days 00:00:00
Name: Order date & time, dtype: object

In [114]:
affected_ids = mismatches["Customer ID"].unique()
gaps = orders[orders["Customer ID"].isin(affected_ids)].groupby("Customer ID")["Order date & time"].agg(lambda x: x.max() - x.min())
gaps.describe()

count                             30
mean               259 days 18:36:56
std      400 days 07:53:28.213926196
min                  0 days 03:00:00
25%                  0 days 22:12:30
50%                  1 days 16:45:00
75%                550 days 21:19:15
max               1338 days 17:40:00
Name: Order date & time, dtype: object

In [ ]:
large_gap_ids = gaps[gaps > pd.Timedelta(days=30)].index
orders[orders["Customer ID"].isin(large_gap_ids)].sort_values(["Customer ID", "Order date & time"])

## Finding: `First order vs repeat` mislabeled for 30 customers (~0.18%)
Comparing the original label against a date-derived version (earliest order per customer = First, rest = Repeat) found 30 customers with at least one wrong label. Root cause unclear — roughly half show orders placed within 1-2 days of each other (likely a pipeline timing issue), the rest show gaps of months to years (possibly a reactivation-relabeling rule, unconfirmed). Given the small scale (<0.2% of customers), the original column is kept as-is for traceability, but a derived `Corrected order type` column is used for all repeat-purchase analysis instead.

In [116]:
catalog[catalog["Cost_per_unit"] >= catalog["MRP"]]

,SKU,Category,Vendor,MRP,Cost_per_unit


In [117]:
customer[["Total orders", "Average order value", "Time to 2nd purchase"]].describe()

,Total orders,Average order value,Time to 2nd purchase
count,16865.000000,16865.000000,7748.000000
mean,1.778832,2528.301471,410.219928
std,1.110665,531.041725,315.313487
min,1.000000,1350.000000,0.000000
25%,1.000000,2150.670000,152.000000
50%,1.000000,2518.000000,342.000000
75%,2.000000,2874.000000,606.000000
max,9.000000,4234.000000,1452.000000


In [118]:
customer.to_csv("../03_cleaned_data/customers_cleaned.csv", index=False)
orders.to_csv("../03_cleaned_data/orders_cleaned.csv", index=False)
order_items.to_csv("../03_cleaned_data/order_line_items_cleaned.csv", index=False)
catalog.to_csv("../03_cleaned_data/sku_catalog_cleaned.csv", index=False)
sessions.to_csv("../03_cleaned_data/website_sessions_cleaned.csv", index=False)

print("Exported files:")
import os
print(os.listdir("../03_cleaned_data"))

Exported files:
['customers_cleaned.csv', 'orders_cleaned.csv', 'order_line_items_cleaned.csv', 'sku_catalog_cleaned.csv', 'website_sessions_cleaned.csv']
